File: tensorrt_edgellm/chat_template.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Copy the checkpoint's chat template into the runtime artifacts."""

from __future__ import annotations

import json
import logging
import shutil
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

__all__ = ["write_chat_template"]

_RUNTIME_ARTIFACTS = ("chat_template.jinja", "chat_template.model",
                      "chat_template.processor")
_UNSUPPORTED_NAMED_TEMPLATE_DIR = "additional_chat_templates"
_LEGACY_INJA_ARTIFACTS = ("chat_template.inja", "chat_template.inja.d")
_PROVIDER_JSON_METADATA = "chat_template.json"
_LEGACY_RUNTIME_JSON_ARTIFACT = "processed_chat_template.json"

# These checkpoints do not publish a Jinja template. Their prompt contracts
# depend on model-specific runtime data, so they use explicit native renderers.
_MANUAL_BY_MODEL_TYPE = {
    "alpamayo_r1": "alpamayo",
    "qwen3_asr": "qwen3_asr",
    "qwen3_asr_thinker": "qwen3_asr",
    "qwen3_tts": "qwen3_tts",
    "qwen3_tts_talker": "qwen3_tts",
    "qwen3_tts_code_predictor": "qwen3_tts",
    "qwen3_tts_code2wav": "qwen3_tts",
}

# These provider processors require message content to be normalized before
# their tokenizer-owned Jinja template is rendered. The C++ runtime implements
# the same documented contract without loading Python or Transformers.
_RAW_PROCESSOR_BY_CLASS = {
    "Phi4MMProcessor": "phi4mm",
}
_RAW_PROCESSOR_BY_MODEL_TYPE = {
    # Legacy InternVL's model.chat() path inserts these media sentinels before
    # applying its text-only tokenizer template.
    "internvl_chat": "internvl",
}


def _read_json(path: Path) -> dict[str, Any]:
    if not path.is_file():
        return {}
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as error:
        raise ValueError(f"failed to read {path}: {error}") from error
    if not isinstance(value, dict):
        raise ValueError(f"{path} must contain a JSON object")
    return value


def _provider_template(value: Any, source: Path) -> str | None:
    """Return the checkpoint's single default provider template."""
    if value is None:
        return None
    if isinstance(value, str):
        if not value.strip():
            raise ValueError(f"{source} contains an empty chat template")
        return value

    if isinstance(value, dict):
        named = value
    elif isinstance(value, list):
        if any(not isinstance(item, dict)
               or not isinstance(item.get("name"), str) or not isinstance(
                   item.get("template"), str) or not item["template"].strip()
               for item in value):
            raise ValueError(
                f"{source} contains an invalid named chat template")
        named = {item["name"]: item["template"] for item in value}
        if len(named) != len(value):
            raise ValueError(
                f"{source} contains duplicate named chat templates")
    else:
        raise ValueError(
            f"{source} has an invalid chat_template value of type "
            f"{type(value).__name__}")

    if set(named) != {"default"}:
        raise ValueError(
            f"{source} defines named chat templates, which are unsupported; "
            "the checkpoint must provide one default template")
    template = named["default"]
    if not isinstance(template, str) or not template.strip():
        raise ValueError(f"{source} contains an invalid default chat template")
    return template


def _embedded_provider_template(model_dir: Path) -> str | None:
    # Transformers may serialize provider-owned Jinja in a JSON metadata
    # container. It is an import source only and is never a runtime artifact.
    for filename in ("chat_template.json", "processor_config.json",
                     "tokenizer_config.json"):
        path = model_dir / filename
        template = _provider_template(
            _read_json(path).get("chat_template"), path)
        if template is not None:
            return template
    return None


def _read_provider_template(path: Path) -> str:
    template = path.read_text(encoding="utf-8")
    if not template.strip():
        raise ValueError(f"{path} contains an empty chat template")
    return template


def _file_provider_template(model_dir: Path) -> Path | None:
    """Return a standalone provider Jinja file when present."""
    default = model_dir / "chat_template.jinja"
    if default.is_file():
        _read_provider_template(default)
    additional = model_dir / _UNSUPPORTED_NAMED_TEMPLATE_DIR
    if additional.exists():
        raise ValueError(
            f"{additional} contains unsupported named chat templates; "
            "the checkpoint must provide one default template")
    return default if default.is_file() else None


def _raw_processor(model_dir: Path) -> str | None:
    processor_classes = set()
    processors = set()
    marker = model_dir / "chat_template.processor"
    if marker.is_file():
        processor = marker.read_text(encoding="utf-8").strip()
        supported = (set(_RAW_PROCESSOR_BY_CLASS.values())
                     | set(_RAW_PROCESSOR_BY_MODEL_TYPE.values()))
        if processor not in supported:
            raise ValueError(f"{marker} names unsupported raw processor "
                             f"{processor!r}")
        processors.add(processor)

    for filename in ("processor_config.json", "preprocessor_config.json",
                     "tokenizer_config.json"):
        config = _read_json(model_dir / filename)
        processor_class = config.get("processor_class")
        if isinstance(processor_class, str):
            processor_classes.add(processor_class)
        auto_map = config.get("auto_map")
        auto_processor = (auto_map.get("AutoProcessor") if isinstance(
            auto_map, dict) else None)
        if isinstance(auto_processor, str):
            processor_classes.add(auto_processor.rsplit(".", 1)[-1])

    processors.update(_RAW_PROCESSOR_BY_CLASS[name]
                      for name in processor_classes
                      if name in _RAW_PROCESSOR_BY_CLASS)
    model_type = _get_model_type(model_dir)
    if model_type in _RAW_PROCESSOR_BY_MODEL_TYPE:
        processors.add(_RAW_PROCESSOR_BY_MODEL_TYPE[model_type])
    if len(processors) > 1:
        raise ValueError(f"checkpoint declares conflicting raw processors: "
                         f"{sorted(processors)}")
    return next(iter(processors), None)


def _get_model_type(model_dir: str | Path) -> str:
    return str(
        _read_json(Path(model_dir) / "config.json").get("model_type", ""))


def _write_raw_processor(output: Path, raw_processor: str | None) -> None:
    if raw_processor is not None:
        (output / "chat_template.processor").write_text(raw_processor + "\n",
                                                        encoding="utf-8")


def _remove_json_runtime_templates(source: Path, output: Path) -> None:
    """Remove JSON copies after materializing the runtime Jinja artifact."""
    (output / _LEGACY_RUNTIME_JSON_ARTIFACT).unlink(missing_ok=True)
    if source.resolve() == output.resolve():
        return

    (output / _PROVIDER_JSON_METADATA).unlink(missing_ok=True)
    for filename in ("processor_config.json", "tokenizer_config.json"):
        path = output / filename
        config = _read_json(path)
        if "chat_template" not in config:
            continue
        del config["chat_template"]
        path.write_text(  # NOSONAR - the artifact filename is fixed above.
            json.dumps(config, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8")


def _clear_runtime_templates(source: Path, output: Path,
                             provider_file: Path | None) -> None:
    preserved = ({provider_file.resolve()}
                 if provider_file is not None else set())
    for stale in _RUNTIME_ARTIFACTS:
        stale_path = output / stale
        if stale_path.exists() and stale_path.resolve() not in preserved:
            stale_path.unlink()

    additional = output / _UNSUPPORTED_NAMED_TEMPLATE_DIR
    source_additional = source / _UNSUPPORTED_NAMED_TEMPLATE_DIR
    copied_additional = (additional.exists() and additional.resolve()
                         != source_additional.resolve())
    if copied_additional:
        if additional.is_dir():
            shutil.rmtree(additional)
        else:
            additional.unlink()

    for stale in _LEGACY_INJA_ARTIFACTS:
        stale_path = output / stale
        if stale_path.is_dir():
            shutil.rmtree(stale_path)
        else:
            stale_path.unlink(missing_ok=True)
    _remove_json_runtime_templates(source, output)  # NOSONAR: fixed names


def write_chat_template(model_dir: str, output_dir: str) -> str:
    """Copy the provider Jinja template for direct Pantor Inja loading."""
    source = Path(model_dir)
    output = Path(output_dir)
    output.mkdir(parents=True, exist_ok=True)

    provider_file = _file_provider_template(source)
    embedded_template = (None if provider_file is not None else
                         _embedded_provider_template(source))
    raw_processor = (_raw_processor(source) if provider_file is not None
                     or embedded_template is not None else None)

    _clear_runtime_templates(source, output, provider_file)

    destination = output / "chat_template.jinja"
    if provider_file is not None:
        if destination.resolve() != provider_file.resolve():
            shutil.copyfile(provider_file, destination)
    elif embedded_template is not None:
        destination.write_text(  # NOSONAR - destination has a fixed filename.
            embedded_template,
            encoding="utf-8")

    if provider_file is not None or embedded_template is not None:
        _write_raw_processor(output, raw_processor)
        logger.info("Copied provider chat template to %s", destination)
        return str(destination)

    model_type = _get_model_type(source)
    manual = _MANUAL_BY_MODEL_TYPE.get(model_type)
    if manual is None:
        raise ValueError(
            f"model_type={model_type!r} does not provide a chat template")

    destination = output / "chat_template.model"
    destination.write_text(manual + "\n", encoding="utf-8")
    logger.info("Wrote native chat renderer %s to %s", manual, destination)
    return str(destination)

File: cpp/chatTemplate/chatTemplate.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "chatTemplate.h"

#include "injaRenderer.h"

#include "common/inputLimits.h"
#include "common/logger.h"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <optional>
#include <stdexcept>
#include <string_view>
#include <unordered_set>
#include <utility>

namespace trt_edgellm::chat_template
{
namespace
{
using Json = nlohmann::json;

enum class RawProcessor
{
    kNone,
    kPhi4MM,
    kInternVL,
    kNemotronOmni,
};

struct MediaCounters
{
    size_t images{0};
    size_t audios{0};
};

std::string trimString(std::string value)
{
    auto const isSpace = [](unsigned char ch) { return std::isspace(ch) != 0; };
    value.erase(value.begin(), std::find_if_not(value.begin(), value.end(), isSpace));
    value.erase(std::find_if_not(value.rbegin(), value.rend(), isSpace).base(), value.end());
    return value;
}

Json parseJsonOrString(std::string const& value)
{
    auto parsed = nlohmann::ordered_json::parse(value, nullptr, false);
    return parsed.is_discarded() ? Json(value) : preserveJsonOrder(parsed);
}

//! Qwen3-Omni Instruct checkpoints are trained never to emit reasoning tokens. Binding
//! ``enable_thinking`` makes their provider template prepend an inert
//! ``<think>\n\n</think>\n\n``, which also shifts the Talker prefill slices off the first
//! spoken token. Leaving the flag undefined selects the no-injection branch, as the
//! provider's own ``apply_chat_template`` does when the caller passes no kwarg.
bool omitsEnableThinking(std::filesystem::path const& modelDir)
{
    auto const configPath = modelDir / "config.json";
    std::ifstream stream(configPath);
    if (!stream)
    {
        return false;
    }
    try
    {
        auto const config = Json::parse(stream);
        auto const model = config.value("model", config.value("model_type", std::string{}));
        return model.rfind("qwen3_omni", 0) == 0;
    }
    catch (std::exception const& error)
    {
        LOG_WARNING("Ignoring unparsable %s: %s", configPath.c_str(), error.what());
        return false;
    }
}

std::string readTextFile(std::filesystem::path const& path)
{
    if (!std::filesystem::is_regular_file(path))
    {
        throw std::runtime_error("file does not exist: " + path.string());
    }
    if (std::filesystem::file_size(path) > limits::tokenizer::kChatTemplateFileSizeBytes)
    {
        throw std::runtime_error("chat template exceeds size limit: " + path.string());
    }
    std::ifstream stream(path);
    if (!stream)
    {
        throw std::runtime_error("failed to open " + path.string());
    }
    return {std::istreambuf_iterator<char>(stream), std::istreambuf_iterator<char>()};
}

std::string textContent(rt::Message const& message)
{
    std::string result;
    for (auto const& item : message.contents)
    {
        if (item.type == "text")
        {
            result += item.content;
        }
    }
    return result;
}

std::string systemContent(rt::Message const& message)
{
    std::string result;
    bool first = true;
    for (auto const& item : message.contents)
    {
        if (item.type != "text")
        {
            continue;
        }
        if (message.contentIsArray && !first)
        {
            result += '\n';
        }
        first = false;
        result += item.content;
    }
    return result;
}

std::vector<rt::Message> normalizeDeveloperMessages(std::vector<rt::Message> messages)
{
    for (auto& message : messages)
    {
        if (message.role == "developer")
        {
            message.role = "system";
        }
    }

    std::vector<std::string> systemContents;
    bool needsConsolidation = false;
    for (size_t index = 0; index < messages.size(); ++index)
    {
        auto const& message = messages[index];
        if (message.role != "system")
        {
            continue;
        }

        needsConsolidation = needsConsolidation || index > 0 || !systemContents.empty();
        auto content = systemContent(message);
        if (!content.empty())
        {
            systemContents.push_back(std::move(content));
        }
    }
    if (!needsConsolidation)
    {
        return messages;
    }

    std::string mergedContent;
    for (size_t index = 0; index < systemContents.size(); ++index)
    {
        if (index > 0)
        {
            mergedContent += "\n\n";
        }
        mergedContent += systemContents[index];
    }
    rt::Message systemMessage;
    systemMessage.role = "system";
    systemMessage.contents.push_back({"text", std::move(mergedContent)});

    std::vector<rt::Message> result;
    result.reserve(messages.size());
    result.push_back(std::move(systemMessage));
    for (auto& message : messages)
    {
        if (message.role != "system")
        {
            result.push_back(std::move(message));
        }
    }
    return result;
}

std::string trajectoryPlaceholder(std::optional<size_t> const& trajectoryPointCount)
{
    if (!trajectoryPointCount)
    {
        throw std::runtime_error("trajectory content requires pastTrajectory data");
    }
    std::string result = rt::kTrajHistoryStartStr;
    for (size_t index = 0; index < 3 * *trajectoryPointCount; ++index)
    {
        result += rt::kTrajHistoryPadStr;
    }
    return result + rt::kTrajHistoryEndStr;
}

std::string phi4MMContent(rt::Message const& message, MediaCounters& counters)
{
    std::string result;
    for (auto const& item : message.contents)
    {
        if (item.type == "text")
        {
            result += item.content;
        }
        else if (item.type == "image")
        {
            result += "<|image_" + std::to_string(++counters.images) + "|>";
        }
        else if (item.type == "audio")
        {
            result += "<|audio_" + std::to_string(++counters.audios) + "|>";
        }
        else
        {
            throw std::runtime_error("Phi4MMProcessor does not accept " + item.type + " message content");
        }
    }
    return result;
}

std::string internVLContent(rt::Message const& message)
{
    std::string result;
    for (auto const& item : message.contents)
    {
        if (item.type == "text")
        {
            result += item.content;
        }
        else if (item.type == "image")
        {
            result += "<IMG_CONTEXT>\n";
        }
        else if (item.type == "video")
        {
            result += "<video>\n";
        }
        else
        {
            throw std::runtime_error("InternVL processor does not accept " + item.type + " message content");
        }
    }
    return result;
}

std::string nemotronOmniContent(rt::Message const& message)
{
    // Inline each media block as its sentinel token (<image>/<audio>/<video>)
    // for the runtime media runner to expand; text blocks pass through.
    std::string result;
    for (auto const& item : message.contents)
    {
        if (item.type == "text")
        {
            result += item.content;
        }
        else if (item.type == "image" || item.type == "audio" || item.type == "video")
        {
            result += "<" + item.type + ">";
        }
        else
        {
            throw std::runtime_error("NemotronOmni processor does not accept " + item.type + " message content");
        }
    }
    return result;
}

void replaceNumberedAliases(std::string& value, std::string_view prefix, std::string_view replacement)
{
    size_t offset = 0;
    while ((offset = value.find(prefix, offset)) != std::string::npos)
    {
        size_t cursor = offset + prefix.size();
        size_t const digitBegin = cursor;
        while (cursor < value.size() && std::isdigit(static_cast<unsigned char>(value[cursor])) != 0)
        {
            ++cursor;
        }
        if (cursor == digitBegin || value.compare(cursor, 2, "|>") != 0)
        {
            offset = cursor;
            continue;
        }
        value.replace(offset, cursor + 2 - offset, replacement);
        offset += replacement.size();
    }
}

std::string applyRawProcessor(std::string value, RawProcessor processor)
{
    if (processor == RawProcessor::kPhi4MM)
    {
        replaceNumberedAliases(value, "<|image_", "<|endoftext10|>");
        replaceNumberedAliases(value, "<|audio_", "<|endoftext11|>");
    }
    return value;
}

Json contentToJson(rt::Message const& message, std::optional<size_t> const& trajectoryPointCount,
    RawProcessor processor, bool expectsContentBlocks, MediaCounters& counters)
{
    if (message.contentIsNull)
    {
        if (!message.contents.empty())
        {
            throw std::runtime_error("null message content cannot contain content blocks");
        }
        return nullptr;
    }
    if (processor == RawProcessor::kPhi4MM)
    {
        return phi4MMContent(message, counters);
    }
    if (processor == RawProcessor::kInternVL)
    {
        return internVLContent(message);
    }
    if (processor == RawProcessor::kNemotronOmni)
    {
        return nemotronOmniContent(message);
    }

    if (!expectsContentBlocks || message.role == "tool")
    {
        // A string-content template renders message.content as a single string.
        // Media requires a rendering contract the model owns: either a provider
        // template that iterates content blocks (expectsContentBlocks), or a
        // raw processor that emits the model's media sentinels. Without one, a
        // media block has no valid placeholder in the string and is rejected.
        std::string content;
        bool first = true;
        for (auto const& item : message.contents)
        {
            if (item.type != "text")
            {
                throw std::runtime_error("provider string-content template requires a model-specific media processor");
            }
            if (!first)
            {
                content += '\n';
            }
            first = false;
            content += item.content;
        }
        return content;
    }

    Json content = Json::array();
    for (auto const& item : message.contents)
    {
        Json block = makeInjaObject({{"type", item.type}});
        if (item.type == "text")
        {
            setInjaObjectField(block, "text", item.content);
        }
        else if (item.type == "trajectory")
        {
            setInjaObjectField(block, "placeholder", trajectoryPlaceholder(trajectoryPointCount));
        }
        else
        {
            setInjaObjectField(block, item.type, item.content);
        }
        content.push_back(std::move(block));
    }
    return content;
}

Json messageToJson(rt::Message const& message, std::optional<size_t> const& trajectoryPointCount,
    RawProcessor processor, bool expectsContentBlocks, MediaCounters& counters)
{
    Json value = makeInjaObject({{"role", message.role}});
    if (message.hasContent)
    {
        setInjaObjectField(
            value, "content", contentToJson(message, trajectoryPointCount, processor, expectsContentBlocks, counters));
    }
    if (message.hasReasoningContent)
    {
        setInjaObjectField(value, "reasoning_content", message.reasoningContent);
        setInjaObjectField(value, "reasoning", message.reasoningContent);
    }
    if (!message.toolCallId.empty())
    {
        setInjaObjectField(value, "tool_call_id", message.toolCallId);
    }
    if (!message.name.empty())
    {
        setInjaObjectField(value, "name", message.name);
    }
    if (message.hasToolCalls || !message.toolCalls.empty())
    {
        Json toolCalls = Json::array();
        for (auto const& call : message.toolCalls)
        {
            Json function = makeInjaObject({{"name", call.name},
                {"arguments", call.argumentsIsString ? Json(call.arguments) : parseJsonOrString(call.arguments)}});
            toolCalls.push_back(
                makeInjaObject({{"id", call.id}, {"type", call.type}, {"function", std::move(function)}}));
        }
        setInjaObjectField(value, "tool_calls", std::move(toolCalls));
    }
    return value;
}

Json toolToJson(rt::ToolDefinition const& tool)
{
    Json function = makeInjaObject({{"name", tool.name}});
    if (tool.hasDescription)
    {
        setInjaObjectField(function, "description", tool.description);
    }
    if (tool.hasParameters)
    {
        setInjaObjectField(function, "parameters", parseJsonOrString(tool.parameters));
    }
    if (tool.hasStrict)
    {
        setInjaObjectField(function, "strict", tool.strict);
    }
    return makeInjaObject({{"type", "function"}, {"function", std::move(function)}});
}

Json toolChoiceToJson(rt::ToolChoice const& toolChoice)
{
    switch (toolChoice.mode)
    {
    case rt::ToolChoice::Mode::kNone: return "none";
    case rt::ToolChoice::Mode::kRequired: return "required";
    case rt::ToolChoice::Mode::kFunction:
        return makeInjaObject(
            {{"type", "function"}, {"function", makeInjaObject({{"name", toolChoice.functionName}})}});
    case rt::ToolChoice::Mode::kAuto: return "auto";
    }
    return "auto";
}

std::string renderRaw(rt::LLMGenerationRequest::Request const& request)
{
    std::string result;
    for (auto const& message : request.messages)
    {
        for (auto const& item : message.contents)
        {
            if (item.type == "text")
            {
                result += item.content;
            }
            else if (item.type == "trajectory")
            {
                auto const count
                    = request.pastTrajectory ? std::optional<size_t>(request.pastTrajectory->size()) : std::nullopt;
                result += trajectoryPlaceholder(count);
            }
            else if (!item.content.empty())
            {
                result += item.content;
            }
            else
            {
                throw std::runtime_error("raw prompt mode requires preformatted content for " + item.type + " inputs");
            }
        }
    }
    return result;
}

std::string renderManual(std::string const& family, std::vector<rt::Message> const& messages,
    std::optional<size_t> const& trajectoryPointCount, bool addGenerationPrompt, bool systemPromptOnly)
{
    std::string result;
    if (family == "alpamayo")
    {
        for (auto const& message : messages)
        {
            result += "<|im_start|>" + message.role + "\n";
            for (auto const& item : message.contents)
            {
                if (item.type == "text")
                    result += item.content;
                else if (item.type == "image")
                    result += "<|vision_start|><|image_pad|><|vision_end|>";
                else if (item.type == "video")
                    result += "<|vision_start|><|video_pad|><|vision_end|>";
                else if (item.type == "trajectory")
                    result += trajectoryPlaceholder(trajectoryPointCount);
                else
                    throw std::runtime_error("alpamayo does not accept " + item.type + " message content");
            }
            result += "<|im_end|>\n";
        }
    }
    else if (family == "qwen3_asr")
    {
        std::string systemPrompt;
        std::string audioPrompt;
        for (auto const& message : messages)
        {
            for (auto const& item : message.contents)
            {
                if (message.role == "system" && item.type == "text")
                    systemPrompt += item.content;
                if (!systemPromptOnly && item.type == "audio")
                    audioPrompt += "<|audio_start|><|audio_pad|><|audio_end|>";
            }
        }
        result = "<|im_start|>system\n" + systemPrompt + "<|im_end|>\n<|im_start|>user\n";
        if (systemPromptOnly)
        {
            return result;
        }
        result += audioPrompt + "<|im_end|>\n";
    }
    else if (family == "qwen3_tts")
    {
        for (auto const& message : messages)
        {
            result += "<|im_start|>" + message.role + "\n";
            for (auto const& item : message.contents)
            {
                if (item.type != "text")
                    throw std::runtime_error("qwen3_tts does not accept " + item.type + " message content");
                result += item.content;
            }
            result += "<|im_end|>\n";
        }
    }
    else
    {
        throw std::runtime_error("unknown native chat renderer: " + family);
    }

    if (addGenerationPrompt)
    {
        result += "<|im_start|>assistant\n";
    }
    return result;
}

} // namespace

class ChatTemplate::Impl
{
public:
    bool load(std::filesystem::path const& modelDir, std::string bosToken, std::string eosToken)
    {
        auto const providerPath = modelDir / "chat_template.jinja";
        auto const additionalPath = modelDir / "additional_chat_templates";
        auto const manualPath = modelDir / "chat_template.model";
        auto const processorPath = modelDir / "chat_template.processor";
        mMode = Mode::kUninitialized;
        mProviderTemplate.reset();
        mManualFamily.clear();
        mRawProcessor = RawProcessor::kNone;
        mOmitEnableThinking = omitsEnableThinking(modelDir);
        mBosToken = std::move(bosToken);
        mEosToken = std::move(eosToken);

        try
        {
            if (std::filesystem::exists(additionalPath))
            {
                throw std::runtime_error(
                    "named provider chat templates are unsupported; the model must provide chat_template.jinja");
            }
            if (std::filesystem::exists(providerPath))
            {
                mProviderTemplate = std::make_unique<InjaRenderer>();
                mProviderTemplate->load(providerPath);
            }

            bool const hasProvider = mProviderTemplate != nullptr;
            bool const hasManual = std::filesystem::exists(manualPath);
            if (hasProvider && hasManual)
            {
                throw std::runtime_error("both provider template and native renderer marker exist");
            }
            if (std::filesystem::exists(processorPath) && !hasProvider)
            {
                throw std::runtime_error("raw processor marker requires a provider template");
            }
            if (hasProvider)
            {
                if (std::filesystem::exists(processorPath))
                {
                    auto const processor = trimString(readTextFile(processorPath));
                    if (processor == "phi4mm")
                    {
                        mRawProcessor = RawProcessor::kPhi4MM;
                    }
                    else if (processor == "internvl")
                    {
                        mRawProcessor = RawProcessor::kInternVL;
                    }
                    else if (processor == "nemotron_omni")
                    {
                        mRawProcessor = RawProcessor::kNemotronOmni;
                    }
                    else
                    {
                        throw std::runtime_error("unsupported raw processor: " + processor);
                    }
                }
                mMode = Mode::kJinja;
                LOG_INFO("Loaded provider Jinja chat template through Pantor Inja from %s", providerPath.c_str());
                return true;
            }
            if (hasManual)
            {
                mManualFamily = trimString(readTextFile(manualPath));
                static std::unordered_set<std::string> const supported{"alpamayo", "qwen3_asr", "qwen3_tts"};
                if (supported.count(mManualFamily) == 0)
                {
                    throw std::runtime_error("unknown native chat renderer: " + mManualFamily);
                }
                mMode = Mode::kManual;
                LOG_INFO("Loaded native chat renderer '%s' from %s", mManualFamily.c_str(), manualPath.c_str());
                return true;
            }
            LOG_ERROR("No provider Jinja or model-specific native chat renderer exists in %s", modelDir.c_str());
        }
        catch (std::exception const& error)
        {
            LOG_ERROR("Failed to load chat template from %s: %s", modelDir.c_str(), error.what());
        }
        return false;
    }

    bool apply(rt::LLMGenerationRequest::Request const& request,
        rt::LLMGenerationRequest::FormattedRequest& formattedRequest, ChatTemplate::Options const& options) const
    {
        if (request.messages.empty())
        {
            LOG_ERROR("Request must contain at least one message");
            return false;
        }

        try
        {
            if (!options.applyTemplate)
            {
                formattedRequest.formattedCompleteRequest = renderRaw(request);
                formattedRequest.formattedSystemPrompt
                    = request.messages.front().role == "system" ? textContent(request.messages.front()) : "";
                return true;
            }
            if (mMode == Mode::kUninitialized)
            {
                throw std::runtime_error("chat template is not loaded");
            }

            auto const* messages = &request.messages;
            std::vector<rt::Message> normalizedMessages;
            auto const trajectoryPointCount
                = request.pastTrajectory ? std::optional<size_t>(request.pastTrajectory->size()) : std::nullopt;
            InjaRenderer const* provider = mMode == Mode::kJinja ? mProviderTemplate.get() : nullptr;
            if (provider != nullptr && !provider->supportsDeveloperRole()
                && std::any_of(request.messages.begin(), request.messages.end(),
                    [](rt::Message const& message) { return message.role == "developer"; }))
            {
                normalizedMessages = normalizeDeveloperMessages(request.messages);
                messages = &normalizedMessages;
            }

            auto render = [&](std::vector<rt::Message> const& messages, bool addGenerationPrompt,
                              bool systemPromptOnly = false) {
                if (mMode == Mode::kManual)
                {
                    return renderManual(
                        mManualFamily, messages, trajectoryPointCount, addGenerationPrompt, systemPromptOnly);
                }

                Json inputs = makeInjaObject({{"messages", Json::array()}});
                MediaCounters counters;
                bool const expectsContentBlocks = provider->expectsContentBlocks();
                for (auto const& message : messages)
                {
                    inputs["messages"].push_back(
                        messageToJson(message, trajectoryPointCount, mRawProcessor, expectsContentBlocks, counters));
                }
                setInjaObjectField(inputs, "tools", options.tools.empty() ? Json(nullptr) : Json::array());
                for (auto const& tool : options.tools)
                {
                    inputs["tools"].push_back(toolToJson(tool));
                }
                setInjaObjectField(inputs, "add_generation_prompt", addGenerationPrompt);
                setInjaObjectField(inputs, "bos_token", mBosToken);
                setInjaObjectField(inputs, "eos_token", mEosToken);
                setInjaObjectField(inputs, "strftime_now", true);
                if (!mOmitEnableThinking)
                {
                    setInjaObjectField(inputs, "enable_thinking", options.enableThinking);
                }
                if (!options.reasoningEffort.empty())
                {
                    setInjaObjectField(inputs, "reasoning_effort", options.reasoningEffort);
                }
                setInjaObjectField(inputs, "parallel_tool_calls", options.parallelToolCalls);
                setInjaObjectField(inputs, "tool_choice", toolChoiceToJson(options.toolChoice));
                setInjaObjectField(inputs, "continue_final_message", false);
                return applyRawProcessor(provider->render(inputs), mRawProcessor);
            };

            formattedRequest.formattedCompleteRequest = render(*messages, options.addGenerationPrompt);
            formattedRequest.formattedSystemPrompt.clear();
            if (messages->front().role == "system")
            {
                if (mMode == Mode::kManual && mManualFamily == "qwen3_asr")
                {
                    formattedRequest.formattedSystemPrompt = render(*messages, false, true);
                }
                else
                {
                    try
                    {
                        auto const prefix = render({messages->front()}, false);
                        if (formattedRequest.formattedCompleteRequest.rfind(prefix, 0) == 0)
                        {
                            formattedRequest.formattedSystemPrompt = prefix;
                        }
                    }
                    catch (std::exception const&)
                    {
                        // Some provider templates require a user turn. In that case,
                        // skip system-only KV reuse instead of inventing a prefix.
                    }
                }
            }
            return true;
        }
        catch (std::exception const& error)
        {
            LOG_ERROR("Failed to apply chat template: %s", error.what());
            return false;
        }
    }

    bool isLoaded() const noexcept
    {
        return mMode != Mode::kUninitialized;
    }

private:
    enum class Mode
    {
        kUninitialized,
        kJinja,
        kManual,
    };

    std::unique_ptr<InjaRenderer> mProviderTemplate;
    std::string mManualFamily;
    std::string mBosToken;
    std::string mEosToken;
    RawProcessor mRawProcessor{RawProcessor::kNone};
    bool mOmitEnableThinking{false};
    Mode mMode{Mode::kUninitialized};
};

ChatTemplate::ChatTemplate()
    : mImpl(std::make_unique<Impl>())
{
}

ChatTemplate::Options ChatTemplate::optionsFrom(rt::LLMGenerationRequest const& request)
{
    Options options;
    options.applyTemplate = request.applyChatTemplate;
    options.addGenerationPrompt = request.addGenerationPrompt;
    options.enableThinking = request.enableThinking;
    options.reasoningEffort = request.reasoningEffort;
    options.tools = request.tools;
    options.toolChoice = request.toolChoice;
    options.parallelToolCalls = request.parallelToolCalls;
    return options;
}

ChatTemplate::~ChatTemplate() = default;
ChatTemplate::ChatTemplate(ChatTemplate&&) noexcept = default;
ChatTemplate& ChatTemplate::operator=(ChatTemplate&&) noexcept = default;

bool ChatTemplate::load(std::filesystem::path const& modelDir, std::string bosToken, std::string eosToken)
{
    return mImpl->load(modelDir, std::move(bosToken), std::move(eosToken));
}

bool ChatTemplate::apply(rt::LLMGenerationRequest::Request const& request,
    rt::LLMGenerationRequest::FormattedRequest& formattedRequest, Options const& options) const
{
    return mImpl->apply(request, formattedRequest, options);
}

bool ChatTemplate::isLoaded() const noexcept
{
    return mImpl->isLoaded();
}

} // namespace trt_edgellm::chat_template
